binding.cc

```diff
@@ -1019,6 +1024,61 @@ namespace zmq {
       MessageReference* msgref_;
   };
 
+  class Socket::OutgoingMessage {
+    public:
+      inline OutgoingMessage(Local<Object> buf) {
+        bufref_ = new BufferReference(buf);
+        if (zmq_msg_init_data(&msg_, Buffer::Data(buf), Buffer::Length(buf),
+            BufferReference::FreeCallback, bufref_) < 0) {
+          delete bufref_;
+          Nan::ThrowError(ErrorMessage());
+        }
+      };
+
+      inline ~OutgoingMessage() {
+        if (zmq_msg_close(&msg_) < 0)
+          Nan::ThrowError(ErrorMessage());
+      };
+
+      inline operator zmq_msg_t*() {
+        return &msg_;
+      }
+
+    private:
+      class BufferReference {
+        public:
+          inline BufferReference(Local<Object> buf) {
+            async = new uv_async_t;
+            uv_async_init(uv_default_loop(), async, Destroy);
```

Comment:
Add error checking, delete `async` and `this` and throw on error, else proceed by setting `async->data`.